cmd/bfgd/bfgd.go

```diff
@@ -148,63 +52,82 @@ func HandleSignals(ctx context.Context, cancel context.CancelFunc, callback func
 
 	select {
 	case <-ctx.Done():
-	case s := <-signalChan: // First signal, cancel context.
+	case s := <-signalChan:
 		if callback != nil {
-			callback(s) // Do whatever caller wants first.
-			cancel()
+			callback(s)
 		}
+		cancel()
 	}
-	<-signalChan // Second signal, hard exit.
+	<-signalChan
+	log.Errorf("Received second termination signal, forcing exit.")
 	os.Exit(2)
 }
 
-func _main() error {
-	// Parse configuration from environment
-	if err := config.Parse(cm); err != nil {
-		return err
-	}
-
-	loggo.ConfigureLoggers(cfg.LogLevel)
-	log.Infof(welcome)
+// initializeLogger configures logging for the application.
+func initializeLogger(logLevel string) {
+	loggo.ConfigureLoggers(logLevel)
+	log.Infof("Logger initialized with level: %s", logLevel)
+}
 
-	pc := config.PrintableConfig(cm)
-	for k := range pc {
-		log.Infof("%v", pc[k])
+// parseConfig parses environment variables into the configuration map.
+func parseConfig() error {
+	if err := config.Parse(cm); err != nil {
+		return fmt.Errorf("failed to parse configuration: %w", err)
 	}
+	return nil
+}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	go HandleSignals(ctx, cancel, func(s os.Signal) {
-		log.Infof("bfg service received signal: %s", s)
-	})
-
+// runServer starts the BFG server with the provided configuration.
+func runServer(ctx context.Context) error {
 	server, err := bfg.NewServer(cfg)
 	if err != nil {
-		return fmt.Errorf("create BFG server: %w", err)
+		return fmt.Errorf("failed to create BFG server: %w", err)
 	}
 	if err = server.Run(ctx); !errors.Is(err, context.Canceled) {
-		return fmt.Errorf("bfg server terminated: %w", err)
+		return fmt.Errorf("server terminated unexpectedly: %w", err)
 	}
-
 	return nil
 }
 
-func init() {
-	version.Component = "bfgd"
-	welcome = "Hemi Bitcoin Finality Governor " + version.BuildInfo()
-}
-
 func main() {
+	// Verify command-line arguments
 	if len(os.Args) != 1 {
-		fmt.Fprintf(os.Stderr, "%v\n", welcome)
-		fmt.Fprintf(os.Stderr, "Usage:\n")
-		fmt.Fprintf(os.Stderr, "\thelp (this help)\n")
-		fmt.Fprintf(os.Stderr, "Environment:\n")
+		fmt.Fprintln(os.Stderr, welcome)
+		fmt.Fprintln(os.Stderr, "Usage:")
+		fmt.Fprintln(os.Stderr, "\thelp (this help)")
+		fmt.Fprintln(os.Stderr, "Environment:")
 		config.Help(os.Stderr, cm)
 		os.Exit(1)
 	}
 
-	if err := _main(); err != nil {
-		log.Errorf("%v", err)
+	// Initialize context for graceful shutdown
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start signal handling in a separate goroutine
+	go handleSignals(ctx, cancel, func(s os.Signal) {
+		log.Infof("Received signal: %s", s)
+	})
+
+	// Parse configuration
+	if err := parseConfig(); err != nil {
+		log.Errorf("Configuration error: %v", err)
+		os.Exit(1)
+	}
+
+	// Initialize logger
+	initializeLogger(cfg.LogLevel)
+
+	// Print welcome message and configuration
+	log.Infof(welcome)
+	pc := config.PrintableConfig(cm)
+	for k, v := range pc {
```

Comment:
What do you think of this change?
```suggestion
	for k, v := range config.PrintableConfig(cm) {
```